src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageDirectiveFeature.cs

```diff
@@ -26,38 +44,36 @@ public static bool TryGetPageDirective(Func<Stream> streamFactory, out string te
                 throw new ArgumentNullException(nameof(streamFactory));
             }
 
-            const string PageDirective = "@page";
-
-            string content;
-            using (var streamReader = new StreamReader(streamFactory()))
+            using (var stream = streamFactory())
             {
-                do
-                {
-                    content = streamReader.ReadLine();
-
-                } while (content != null && string.IsNullOrWhiteSpace(content));
-                content = content?.Trim();
+                var sourceDocument = RazorSourceDocument.ReadFrom(stream, fileName: "Parse.cshtml");
+                return TryGetPageDirective(sourceDocument, out template);
             }
+        }
+
+        private static bool TryGetPageDirective(RazorSourceDocument sourceDocument, out string template)
+        {
+            var codeDocument = RazorCodeDocument.Create(sourceDocument);
+            PageDirectiveEngine.Process(codeDocument);
 
-            if (content == null || !content.StartsWith(PageDirective, StringComparison.Ordinal))
+            if (PageDirective.TryGetPageDirective(codeDocument.GetIRDocument(), out var pageDirective))
             {
-                template = null;
-                return false;
+                template = pageDirective.RouteTemplate;
+                return true;
             }
 
-            template = content.Substring(PageDirective.Length, content.Length - PageDirective.Length).TrimStart();
+            template = null;
+            return false;
+        }
 
-            if (template.StartsWith("\"") && template.EndsWith("\""))
-            {
-                template = template.Substring(1, template.Length - 2);
-            }
-            // If it's not in quotes it's not our template
-            else
+        private class PageDirectiveParserOptionsFeature : RazorEngineFeatureBase, IRazorParserOptionsFeature
+        {
+            public int Order { get; }
+
+            public void Configure(RazorParserOptionsBuilder options)
             {
-                template = string.Empty;
+                options.ParseOnlyLeadingDirectives = true;
```

Comment:
Unrelated to this change. @rynowak, I somehow missed this rename. Shouldn't this be `ParseOnlyLeadingDirective`? We only parse the first directive, not multiple.